src/dev/so_migration/snapshot_plugin_types.ts

```diff
@@ -0,0 +1,152 @@
+  // eslint-disable-next-line @kbn/imports/no_boundary_crossing
+} from '@kbn/core-test-helpers-kbn-server';
+import { REPO_ROOT } from '@kbn/repo-info';
+import { ToolingLog } from '@kbn/tooling-log';
+
+import { mkdirp } from '../build/lib';
+
+type MigrationInfoRecord = Pick<
+  SavedObjectTypeMigrationInfo,
+  'name' | 'migrationVersions' | 'schemaVersions' | 'modelVersions'
+> & {
+  hash: string;
+};
+
+type ServerHandles = Awaited<ReturnType<typeof startServers>> | undefined;
+
+/**
+ * Starts up ES & Kibana to extract plugin migration information, and save it to @param outputPath.
+ * @param log Tooling log handed over from CLI runner
+ * @param outputPath Path (absolute or relative to) where the extracted migration info should be saved to in a JSON format.
+ */
+async function takeSnapshot({ log, outputPath }: { log: ToolingLog; outputPath: string }) {
+  let serverHandles: ServerHandles;
+
+  const snapshotOutputPath = path.isAbsolute(outputPath)
+    ? outputPath
+    : path.resolve(REPO_ROOT, outputPath);
+
+  try {
+    serverHandles = await startServers();
+
+    const typeRegistry = serverHandles.coreStart.savedObjects.getTypeRegistry();
+    const allTypes = typeRegistry.getAllTypes();
+
+    const migrationInfoMap = allTypes.reduce((map, type) => {
+      const migrationInfo = extractMigrationInfo(type);
+      map[type.name] = {
+        name: migrationInfo.name,
+        migrationVersions: migrationInfo.migrationVersions,
+        hash: getMigrationHash(type),
+        modelVersions: migrationInfo.modelVersions,
+        schemaVersions: migrationInfo.schemaVersions,
+      };
+      return map;
+    }, {} as Record<string, MigrationInfoRecord>);
+
+    await writeSnapshotFile(snapshotOutputPath, migrationInfoMap);
+    log.info('Snapshot taken!');
+
+    return migrationInfoMap;
+  } finally {
+    log.debug('Shutting down servers');
+    await shutdown(log, serverHandles);
+  }
+}
+
+async function startServers() {
+  const { startES } = createTestServers({
+    adjustTimeout: () => {},
+  });
+
+  const esServer = await startES();
+  const kibanaRoot = createRootWithCorePlugins({}, { oss: false });
+  await kibanaRoot.preboot();
+  await kibanaRoot.setup();
+  const coreStart = await kibanaRoot.start();
+  return { esServer, kibanaRoot, coreStart };
+}
+
+async function writeSnapshotFile(
+  snapshotOutputPath: string,
+  hashMap: Record<string, MigrationInfoRecord>
+) {
+  const timestamp = Date.now();
+  const date = new Date().toISOString();
+  const buildUrl = process.env.BUILDKITE_BUILD_URL;
+  const prId = process.env.BUILDKITE_MESSAGE?.match(/\(#(\d+)\)/)?.[1];
+  const pullRequestUrl = prId ? `https://github.com/elastic/kibana/pulls/${prId}` : null;
+  const kibanaCommitHash = process.env.BUILDKITE_COMMIT || getLocalHash();
+
+  const payload = {
+    meta: {
+      timestamp,
+      date,
+      kibanaCommitHash,
+      buildUrl,
+      pullRequestUrl,
+    },
+    typeHashes: hashMap,
```

Comment:
Similar NIT: `typeHashes` should probably be `typeDefinitions`, `typeInfos` or something?

Reply:
good catches, these are just leftovers from the function extracts - I'll find better names